### PR TITLE
Fixed building UWP app using Unity 2019

### PR DESCRIPTION
**Pull request by AndreiShilkin**
Fixed `Debugger.cpp` generation logic.
[AB#61609](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/61609)
#259

### CODE REVIEW

**Review by annakocheshkova — DISMISSED** [review]

**Review by dtrabo — DISMISSED** [review]

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 1]
```diff
@@ -178,10 +179,93 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+        var codeLines = File.ReadAllLines(destDebuggerPath).ToList();
+
+        // Update #include and #undef derictives.
+        var lastIncludeLineIndex = SearchForLine(codeLines, "#include", true);
+        if (lastIncludeLineIndex == -1)
+        {
+            throw new Exception("Unexpected content of Debugger.cpp");
+        }
+        codeLines.Insert(lastIncludeLineIndex + 1, "#include <Windows.h>");
+        codeLines.Insert(lastIncludeLineIndex + 2, "#undef GetCurrentDirectory");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs:209]: Please add comment WHY it works

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs (hunk 1 above)
Comment (AndreiShilkin) [reply Assets/AppCenter/Editor/AppCenterPostBuild.cs:209]: Done

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 2]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+
+        var codeLines = File.ReadAllLines(destDebuggerPath).ToList();
+
+        // Update #include and #undef derictives.
+        var lastIncludeLineIndex = SearchForLine(codeLines, "#include", true);
+        if (lastIncludeLineIndex == -1)
+        {
+            throw new Exception("Unexpected content of Debugger.cpp");
+        }
+        codeLines.Insert(lastIncludeLineIndex + 1, "#include <Windows.h>");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs:203]:
```suggestion
        // Add '#include <Windows.h>' which provides 'OutputDebugStringW'.
        codeLines.Insert(lastIncludeLineIndex + 1, "#include <Windows.h>");
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 3]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+        codeLines.Insert(lastIncludeLineIndex + 2, "#undef GetCurrentDirectory");
+        // Add logging method.
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs:211]:
```suggestion

        // Add logging method.
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 4]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+        var codeLines = File.ReadAllLines(destDebuggerPath).ToList();
+
+        // Update #include and #undef derictives.
+        var lastIncludeLineIndex = SearchForLine(codeLines, "#include", true);
+        if (lastIncludeLineIndex == -1)
+        {
+            throw new Exception("Unexpected content of Debugger.cpp");
+        }
+        codeLines.Insert(lastIncludeLineIndex + 1, "#include <Windows.h>");
+        codeLines.Insert(lastIncludeLineIndex + 2, "#undef GetCurrentDirectory");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs:209]:
```suggestion
        /*
         * 'GetCurrentDirectory' define conflicts with generated code and new versions of Unity
         * combine some files on the compilation so the changes in one file can affect another.
         */
        codeLines.Insert(lastIncludeLineIndex + 2, "#undef GetCurrentDirectory");
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 5]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+        // Add logging method.
+        var logMethodLineIndex = SearchForLine(codeLines, "void Debugger::Log");
+        if (logMethodLineIndex == -1)
+        {
+            throw new Exception("Unexpected content of Debugger.cpp");
+        }
+        var insertingPosition = GetFirstLineInMethodBody(codeLines, logMethodLineIndex);
+        codeLines.Insert(insertingPosition, "OutputDebugStringW(message->chars);");
+        // Enable logging.
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs:220]:
```suggestion

        // Enable logging.
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 6]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
+        var isLoggingMethodLineIndex = SearchForLine(codeLines, "bool Debugger::IsLogging");
+        if (isLoggingMethodLineIndex == -1)
+        {
+            throw new Exception("Unexpected content of Debugger.cpp");
+        }
+        var firstLineInMethodBody = GetFirstLineInMethodBody(codeLines, isLoggingMethodLineIndex);
+        var lastLineInMethodBody = GetLastLineInMethodBody(codeLines, isLoggingMethodLineIndex);
+        codeLines.RemoveRange(firstLineInMethodBody, lastLineInMethodBody - firstLineInMethodBody);
+        codeLines.Insert(firstLineInMethodBody, "return true;");
+
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs]:
```suggestion
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 7]
```diff
@@ -176,12 +177,103 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
...
     {
-        var sourceDebuggerPath = "Assets\\AppCenter\\Plugins\\WSA\\IL2CPP\\Debugger.cpp.txt";
         var destDebuggerPath = Path.Combine(pathToBuiltProject,
             "Il2CppOutputProject\\IL2CPP\\libil2cpp\\icalls\\mscorlib\\System.Diagnostics\\Debugger.cpp");
-        File.Copy(sourceDebuggerPath, destDebuggerPath, true);
+        if (!File.Exists(destDebuggerPath))
+        {
+            throw new FileNotFoundException("Debugger.cpp file not found");
+        }
+
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs]:
```suggestion
```

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 8]
```diff
@@ -176,12 +177,106 @@ public static void InjectCodeToFile(string appFilePath, string appFileName, stri
         }
     }
 
+    /**
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs]: Use C# xml-doc

**Review by MatkovIvan — DISMISSED** [review]

**Comment by AndreiShilkin** [comment]
_Notes:_
We have to use `WriteAllLines(String, String[])` method overloading to have backward compatibility with Unity 2017. [Link to docs.](https://docs.microsoft.com/ru-ru/dotnet/api/system.io.file.writealllines?view=netframework-3.5)

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**